File: CPlusPlus/611. 有效三角形的个数.cpp
/*
 * 暴力解法， O(n^3)
 * 划定窗口实现 o(n^2)： 因为 mid 和 right 可以朝着一个方向挪动
 * 二分查找 O(n^2 log n) 
*/
#include <vector>
#include <algorithm>

using namespace std;
class BruteForce_Solution {
public:
    int triangleNumber(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int beg = 0; 
        int end = nums.size()-1;
        int sum = 0;
        while(end > beg+1){
            int inner_beg = beg;
            while(inner_beg < end-1){
                int mid = inner_beg+1;
                while(mid < end && nums[inner_beg] + nums[mid] <= nums[end]){
                    ++mid;
                }
                sum += (end -mid);
                ++inner_beg;
            }
            --end;
        }
        return sum;
    }
};

class SlidingWindow_Solution {
public:
    int triangleNumber(vector<int>& nums) {
        if(nums.size() < 3) return 0;
        sort(nums.begin(), nums.end());
        int res{0};
        for(int i = 0; i < nums.size()-2; ++i){
            int mid = i+1;
            int right = mid + 1;
            while(right < nums.size()){
                while(mid < right && nums[i] + nums[mid] <= nums[right]){
                    ++mid;
                }
                res += right-mid;
                ++right;
            }
        }
        return res;
    }
};

class BinarySearch_Solution {
public:
    int triangleNumber(vector<int>& nums) {
        if(nums.size() < 3) return 0;
        sort(nums.begin(), nums.end());
        int res{0};
        int n = nums.size();
        for(int i = 0; i < nums.size()-2; ++i){
            for(int j = i+1; j < nums.size()-1; ++j){
                int k = binary_search(nums, nums[i]+nums[j]);
                if(k>j) res += (k - j-1);
            }
        }
        return res;
    }
private:
    int binary_search(vector<int>& nums, int val){
        int l = 0;
        int r = nums.size();
        while(l < r){
            int mid = (l + r)>>1;
            if(nums[mid] >= val){
                r = mid;
            }else{
                l = mid + 1;
            }
        }
        return l;
    }
};

File: Rust/611. 有效三角形的个数.rs
// 使用指针挪动，固定 low, 然后挪动 mid 和 high
// 这里需要注意的是， mid 和 high 往一个方向挪动， 时间复杂度依旧是是 O(n)
// 因为 mid 最多挪动 n 次， high 也最多挪动 n 次， 并且两者的挪动不是 nested
pub fn triangle_number(mut nums: Vec<i32>) -> i32 {
    if nums.len() < 3 {
        return 0;
    }
    nums.sort();
    let mut low: usize = 0;
    let mut res: i32 = 0;
    while low < nums.len() - 2 {
        let mut mid = low + 1;
        let mut high = mid + 1;
        while mid < high || high < nums.len() {
            while high < nums.len() && (high <= mid || nums[low] + nums[mid] > nums[high]) {
                high = high + 1;
            }
            res = res + (high - mid - 1) as i32;
            mid = mid + 1;
        }
        low = low + 1;
    }
    res
}
